elkscmd/man/man1/ps.1
.TH ps 1
.SH NAME
ps \- list processes or uptime status
.SH SYNOPSIS
.B ps
[\fB-lu\fR]
.SH DESCRIPTION
.BR ps
prints a list of all the processes currently running on the system, giving
a line for each process. The process id, process group, tty, user, state,
CPU percentage utilization,
process heap total and free, overall process memory size, and the process name
and command line are displayed.
The kernel code an data segment address of the corresponding binary are also
displayed when the
.B -l
option is used.
The
.B uptime
command is linked to
.BR ps.
.PP
Each line of the listing contains:
.TP 10
PID
The process ID.
.TP 10
GRP
The process group.
.TP 10
TTY
The controlling TTY.
.TP 10
USER
The user ID.
.TP 10
STAT
Process status (see below).
.TP 10
CPU
CPU % utilization.
.TP 10
CSEG
Code segment in main memory in hex.
.TP 10
DSEG
Data segment in main memory in hex.
.TP 10
HEAP
Current heap used (decimal).
.TP 10
FREE
Remaining heap and stack space available (decimal).
.TP 10
SIZE
Total main memory used in bytes.
.TP 10
COMMAND
Program name and arguments.
.SS OPTIONS
.TP
.B \-l
Display DSEG and CSEG columns.
.TP
.B \-u
Display system uptime.
.SH "PROCESS STATES"
Process states are from the following table:
.TP 10
R
Running
.TP 10
S
Interruptible sleeping
.TP 10
s
Uninterruptible sleeping
.TP 10
T
Stopped
.TP 10
Z
Zombie
.TP 10
E
Exiting
.SH EXIT STATUS
.TP
.I 0
Process table read successfully.
.TP
.I 1
Unable to access 
.I /dev/kmem
or could not read task entries.
.SH FILES
.PD 0
.TP
.B /dev/kmem
.sp
.SH AUTHORS
Al Riddoch (ajr@ecs.soton.ac.uk)
.br
Greg Haerr (greg@censoft.com)


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/sys_utils/ps.c
/*
 * ps.c
 * Copyright 1998 Alistair Riddoch
 * ajr@ecs.soton.ac.uk
 *
 * This file may be distributed under the terms of the GNU General Public
 * License v2, or at your option any later version.
 *
 * This is a small version of ps for use in the ELKS project.
 * Enhanced by Greg Haerr 17 Apr 2020
 */
#define __KERNEL__
#include <linuxmt/ntty.h>
#undef __KERNEL__

#include <autoconf.h>           /* for CONFIG_ options */
#include <linuxmt/mm.h>
#include <linuxmt/mem.h>
#include <linuxmt/major.h>
#include <linuxmt/sched.h>
#include <linuxmt/fixedpt.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <string.h>
#include <dirent.h>
#include <pwd.h>
#include <getopt.h>

#define LINEARADDRESS(off, seg)		((off_t) (((off_t)seg << 4) + off))

int memread(int fd, word_t off, word_t seg, void *buf, int size)
{
	if (lseek(fd, LINEARADDRESS(off, seg), SEEK_SET) == -1)
		return 0;

	if (read(fd, buf, size) != size)
		return 0;

	return 1;
}

word_t getword(int fd, word_t off, word_t seg)
{
	word_t word;

	if (!memread(fd, off, seg, &word, sizeof(word)))
		return 0;
	return word;
}

void process_name(int fd, unsigned int off, unsigned int seg)
{
	word_t argc, argv;
	char buf[80];

	argc = getword(fd, off, seg);

	while (argc-- > 0) {
		off += 2;
		argv = getword(fd, off, seg);
		if (!memread(fd, argv, seg, buf, sizeof(buf)))
			return;
		printf("%s ",buf);
	}
}


char *devname(unsigned int minor)
{
	struct dirent *d;
	dev_t ttydev = MKDEV(TTY_MAJOR, minor);
	struct stat st;
	static char dev[] = "/dev";
	static char name[MAXNAMLEN+1];

	DIR *fp = opendir(dev);
	if (fp == 0)
		return "??";
	strcpy(name, dev);
	strcat(name, "/");

	while ((d = readdir(fp)) != 0) {
		if (strlen(d->d_name) > sizeof(name) - sizeof(dev) - 1)
			continue;
		if (d->d_name[0] == '.')
			continue;
		strcpy(name + sizeof(dev), d->d_name);
		if (!stat(name, &st) && st.st_rdev == ttydev) {
			closedir(fp);
			return name+8;
		}
	}
	closedir(fp);
	return "?";
}

char *tty_name(int fd, unsigned int off, unsigned int seg)
{
	off_t addr = ((off_t)seg << 4) + off;
	struct tty tty;

	if (off == 0)
		return "";
	if (lseek(fd, addr, SEEK_SET) == -1) return "?";

	if (read(fd, &tty, sizeof(tty)) != sizeof(tty)) return "?";

	return devname(tty.minor);
}

int main(int argc, char **argv)
{
	int c, fd;
	unsigned int j, ds, off;
	word_t cseg, dseg;
	struct task_struct task_table;
	struct passwd * pwent;
    int f_listall = 0;
    char *progname = argv[0];
    int f_uptime = !strcmp(progname, "uptime");

    while ((c = getopt(argc, argv, "lu")) != -1) {
        switch (c) {
        case 'l':       /* list all - CSEG/DSEG */
            f_listall = 1;
            break;
        case 'u':       /* uptime */
            f_uptime = 1;
            break;
        default:
            printf("Usage: %s: [-lu]\n", progname);
            exit(1);
        }
    }

	if ((fd = open("/dev/kmem", O_RDONLY)) < 0) {
		perror("ps");
		exit(1);
	}
	if (ioctl(fd, MEM_GETDS, &ds) < 0) {
		perror("ps");
		exit(1);
	}

#ifdef CONFIG_CPU_USAGE
    if (f_uptime) {
        jiff_t uptime;
        unsigned int upoff;

	    if (ioctl(fd, MEM_GETUPTIME, &upoff) < 0 ||
                !memread(fd, upoff, ds, &uptime, sizeof(uptime))) {
		    perror("ps");
		    exit(1);
	    }

        unsigned long n = uptime / HZ;
        int days = n / (24 * 3600L);
        n = n % (24 * 3600L);
        int hours = n / 3600L;
        n %= 3600;
        int minutes = n / 60 ;

        printf("up for %d days, %d hour%s, and %d minute%s\n",
            days, hours, hours == 1? "": "s", minutes, minutes == 1? "": "s");
        exit(0);
    }
#endif

	if (ioctl(fd, MEM_GETTASK, &off) < 0) {
		perror("ps");
		exit(1);
	}

	printf("  PID   GRP  TTY USER STAT ");
#ifdef CONFIG_CPU_USAGE
    printf("CPU");
#endif
    printf(" ");
	if (f_listall) printf("CSEG DSEG ");
	printf(" HEAP  FREE   SIZE COMMAND\n");
	for (j = 1; j <= MAX_TASKS; j++) {
		if (!memread(fd, off + j*sizeof(struct task_struct), ds, &task_table, sizeof(task_table))) {
			perror("ps");
			return 1;
		}

		if (task_table.t_kstackm != KSTACK_MAGIC)
			break;
		if (task_table.t_regs.ss == 0)
			continue;

		switch (task_table.state) {
		case TASK_UNUSED:			continue;
		case TASK_RUNNING:			c = 'R'; break;
		case TASK_INTERRUPTIBLE:	c = 'S'; break;
		case TASK_UNINTERRUPTIBLE:	c = 's'; break;
		case TASK_STOPPED:			c = 'T'; break;
		case TASK_ZOMBIE:			c = 'Z'; break;
		case TASK_EXITING:			c = 'E'; break;
		default:					c = '?'; break;
		}
		pwent = (getpwuid(task_table.uid));

		/* pid grp tty user stat*/
		printf("%5d %5d %4s %-8s%c ",
				task_table.pid,
				task_table.pgrp,
				tty_name(fd, (unsigned int)task_table.tty, ds),
				(pwent ? pwent->pw_name : "unknown"),
				c);

#ifdef CONFIG_CPU_USAGE
        {
            /* Round up, then divide by 2 for %. Change if SAMP_FREQ not 2 */
            unsigned long cpu_percent = (task_table.average + FIXED_HALF) >> 1;
            printf("%3d", FIXED_INT(cpu_percent));
        }
#endif
		/* CSEG*/
		cseg = (word_t)task_table.mm.seg_code;
		if (f_listall) printf(" %4x ",
            cseg? getword(fd, (word_t)cseg+offsetof(struct segment, base), ds): 0);

		/* DSEG*/
		dseg = (word_t)task_table.mm.seg_data;
		if (f_listall) printf("%4x",
            dseg? getword(fd, (word_t)dseg+offsetof(struct segment, base), ds): 0);

		if (dseg) {
			/* heap*/
			printf(" %5u ", (word_t)(task_table.t_endbrk - task_table.t_enddata));

			/* free*/
			printf("%5u ", (word_t)(task_table.t_regs.sp - task_table.t_endbrk));

			/* stack*/
			//printf("%5u ", (word_t)(task_table.t_begstack - task_table.t_regs.sp));

			/* size*/
			segext_t size = getword(fd, (word_t)cseg+offsetof(struct segment, size), ds)
							+ getword(fd, (word_t)dseg+offsetof(struct segment, size), ds);
			printf("%6ld ", (long)size << 4);

			process_name(fd, task_table.t_begstack, task_table.t_regs.ss);
		}
		printf("\n");
	}
	return 0;
}
